### PR TITLE
4. test(db): add large transaction tests

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

These tests check that the smaller keys in #3152 don't break large transaction storage.

## Solution

- generate a test block with many transparent outputs
- add block and transaction round-trip tests
- fix large block serialization instability in the tests

Associated refactors:
- split out a block data write method
- add a height argument to new_outputs

## Review

This PR is based on PR #3741, it should rebase automatically.

### Reviewer Checklist

  - [ ] Code refactors are reasonable
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Actually make the changes:
- #3152 
- #3151

### CODE REVIEW

**Comment by jvff** [comment]
>     * add a height argument to new_outputs

Could you explain a bit more why the `height` argument was added? Feels risky to be able to have a `block` and a `height` parameter where the `block` could have a height different from the `height` parameter.

**Comment by teor2345** [comment]
> > ```
> > * add a height argument to new_outputs
> > ```
> 
> Could you explain a bit more why the `height` argument was added? Feels risky to be able to have a `block` and a `height` parameter where the `block` could have a height different from the `height` parameter.

This is a temporary fix for two issues:
- the generated blocks don't have a coinbase height - I'll add a TODO
- Zebra accepts blocks without coinbase heights at parse time - #2386 

I didn't want to fix the generated blocks, because that's out of scope. But I can create test-only methods, if that helps?

**Comment by teor2345** [comment]
> > > ```
> > > * add a height argument to new_outputs
> > > ```
> > 
> > 
> > Could you explain a bit more why the `height` argument was added? Feels risky to be able to have a `block` and a `height` parameter where the `block` could have a height different from the `height` parameter.
> 
>     * the generated blocks don't have a coinbase height - I'll add a TODO
> 
> 
> I didn't want to fix the generated blocks, because that's out of scope. But I can create test-only methods, if that helps?

Fixed in 1db872801.

File: zebra-chain/src/block/tests/generate.rs [hunk 1]
```diff
@@ -130,3 +156,64 @@ fn single_transaction_block(oversized: bool) -> Block {
...
+        outputs,
+        lock_time,
+    };
+
+    // Put the big transaction into a block
+    let transactions = vec![Arc::new(big_transaction)];
+    Block {
+        header,
+        transactions,
+    }
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs]: Idea: should we add an assertion here? Something like:

```suggestion
    let block = Block {
        header,
        transactions,
    };

    assert_eq!(oversized, block.zcash_serialize_to_vec().len() > MAX_BLOCK_BYTES);
    
    block
```

**Review by jvff — DISMISSED** [review]
Looks good! I added one optional idea, but we can merge as is :+1:

**Comment by teor2345** [comment]
> Looks good! I added one optional idea, but we can merge as is 👍

Happy to do this if I get time.

File: zebra-chain/src/block/tests/generate.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/block/tests/generate.rs]: Good catch!

Fixed in [fe033cf](https://github.com/ZcashFoundation/zebra/pull/3759/commits/fe033cf255508dcaeec206fd9aacfcb4ac8d1988).

The under-sized blocks were far too small, so this required a significant workaround for the test vectors, and multiple fixes to the size calculations.

I'm happy with the workarounds now, hope you are too!

**Review by conradoplg — APPROVED** [review]
Reapproving after rebase

File: zebra-chain/src/block/tests/generate.rs [hunk 2]
```diff
@@ -1,106 +1,207 @@
...
+    );
+    assert!(
+        serialized_len > MIN_LARGE_BLOCK_BYTES.try_into().unwrap(),
+        "block is large\n\
+         oversized: {},\n\
+         serialized_len: {},\n\
+         MIN_LARGE_BLOCK_BYTES: {},",
+        oversized,
+        serialized_len,
+        MIN_LARGE_BLOCK_BYTES,
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs:175]: Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MIN_LARGE_BLOCK_BYTES: {MIN_LARGE_BLOCK_BYTES},",
```

File: zebra-chain/src/block/tests/generate.rs [hunk 3]
```diff
@@ -1,106 +1,207 @@
...
+    assert_eq!(
+        oversized,
+        serialized_len > MAX_BLOCK_BYTES.try_into().unwrap(),
+        "block is over-sized if requested:\n\
+         oversized: {},\n\
+         serialized_len: {},\n\
+         MAX_BLOCK_BYTES: {},",
+        oversized,
+        serialized_len,
+        MAX_BLOCK_BYTES,
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs:165]: Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MAX_BLOCK_BYTES: {MAX_BLOCK_BYTES},",
```

File: zebra-chain/src/block/tests/generate.rs [hunk 4]
```diff
@@ -125,8 +226,112 @@ fn single_transaction_block(oversized: bool) -> Block {
...
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs:245]: Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MAX_BLOCK_BYTES: {MAX_BLOCK_BYTES},",
```

File: zebra-chain/src/block/tests/generate.rs [hunk 5]
```diff
@@ -125,8 +226,112 @@ fn single_transaction_block(oversized: bool) -> Block {
...
+    );
+    assert!(
+        serialized_len > MIN_LARGE_BLOCK_BYTES.try_into().unwrap(),
+        "block is large\n\
+         oversized: {},\n\
+         serialized_len: {},\n\
+         MIN_LARGE_BLOCK_BYTES: {},",
+        oversized,
+        serialized_len,
+        MIN_LARGE_BLOCK_BYTES,
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs:255]: Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MIN_LARGE_BLOCK_BYTES: {MIN_LARGE_BLOCK_BYTES},",
```

File: zebra-chain/src/block/tests/generate.rs [hunk 6]
```diff
@@ -125,8 +226,112 @@ fn single_transaction_block(oversized: bool) -> Block {
...
+    assert_eq!(
+        oversized,
+        serialized_len > MAX_BLOCK_BYTES.try_into().unwrap(),
+        "block is over-sized if requested:\n\
+         oversized: {},\n\
+         serialized_len: {},\n\
+         MAX_BLOCK_BYTES: {},",
+        oversized,
+        serialized_len,
+        MAX_BLOCK_BYTES,
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs:323]: Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MAX_BLOCK_BYTES: {MAX_BLOCK_BYTES},",
```

File: zebra-chain/src/block/tests/generate.rs [hunk 7]
```diff
@@ -125,8 +226,112 @@ fn single_transaction_block(oversized: bool) -> Block {
...
+    );
+    assert!(
+        serialized_len > MIN_LARGE_BLOCK_BYTES.try_into().unwrap(),
+        "block is large\n\
+         oversized: {},\n\
+         serialized_len: {},\n\
+         MIN_LARGE_BLOCK_BYTES: {},",
+        oversized,
+        serialized_len,
+        MIN_LARGE_BLOCK_BYTES,
```
Comment (jvff) [inline zebra-chain/src/block/tests/generate.rs:333]: Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MIN_LARGE_BLOCK_BYTES: {MIN_LARGE_BLOCK_BYTES},",
```

**Review by jvff — COMMENTED** [review]
Sorry for adding another late suggestion, feel free to ignore it and merge as is if you want to :+1:

**Review by jvff — APPROVED** [review]

**Comment by teor2345** [comment]
@mergifyio refresh

File: zebra-chain/src/block/tests/generate.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/block/tests/generate.rs:175]: Sorry, I think it's a bit late for formatting changes on this PR.

Happy to think about this in future though.

**Pull request merged**